From 756612fcd3a945ac59d52920bb77ba1302a13a5f Mon Sep 17 00:00:00 2001
From: Fei Wang <fei.w.wang-at-intel.com@ffmpeg.org>
Date: Fri, 29 Apr 2022 15:59:38 +0800
Subject: [PATCH] lavu: add sub frame side data

Sub frame side data allows attach another AVFrame as side data into
the target AVFrame.

Note: remove version bump in libavutil/version.h compare to version
submitted to community:

https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=6504

Signed-off-by: Fei Wang <fei.w.wang@intel.com>
---
 libavutil/Makefile             |  4 ++-
 libavutil/frame.c              |  5 ++-
 libavutil/frame.h              |  5 +++
 libavutil/sub_frame_metadata.c | 61 ++++++++++++++++++++++++++++++++++
 libavutil/sub_frame_metadata.h | 35 +++++++++++++++++++
 5 files changed, 108 insertions(+), 2 deletions(-)
 create mode 100644 libavutil/sub_frame_metadata.c
 create mode 100644 libavutil/sub_frame_metadata.h

diff --git a/libavutil/Makefile b/libavutil/Makefile
index 6e6fa8d800..1ec25a9d1e 100644
--- a/libavutil/Makefile
+++ b/libavutil/Makefile
@@ -94,7 +94,8 @@ HEADERS = adler32.h                                                     \
           tea.h                                                         \
           tx.h                                                          \
           film_grain_params.h                                           \
-          video_hint.h
+          video_hint.h							\
+          sub_frame_metadata.h                                          \
 
 ARCH_HEADERS = bswap.h                                                  \
                intmath.h                                                \
@@ -190,6 +191,7 @@ OBJS = adler32.o                                                        \
        video_enc_params.o                                               \
        video_hint.o                                                     \
        film_grain_params.o                                              \
+       sub_frame_metadata.o                                             \
 
 
 OBJS-$(CONFIG_CUDA)                     += hwcontext_cuda.o
diff --git a/libavutil/frame.c b/libavutil/frame.c
index 930f01dc19..a48501a43d 100644
--- a/libavutil/frame.c
+++ b/libavutil/frame.c
@@ -56,6 +56,7 @@ static const AVSideDataDescriptor sd_props[] = {
     [AV_FRAME_DATA_SPHERICAL]                   = { "Spherical Mapping",                            AV_SIDE_DATA_PROP_GLOBAL },
     [AV_FRAME_DATA_ICC_PROFILE]                 = { "ICC profile",                                  AV_SIDE_DATA_PROP_GLOBAL },
     [AV_FRAME_DATA_SEI_UNREGISTERED]            = { "H.26[45] User Data Unregistered SEI message",  AV_SIDE_DATA_PROP_MULTI },
+    [AV_FRAME_DATA_SUB_FRAME]                   = { "Sub frame Metadata" },
 };
 
 static void get_frame_defaults(AVFrame *frame)
@@ -356,7 +357,9 @@ FF_ENABLE_DEPRECATION_WARNINGS
         if (   sd_src->type == AV_FRAME_DATA_PANSCAN
             && (src->width != dst->width || src->height != dst->height))
             continue;
-        if (force_copy) {
+        /* Don't copy sub frame side data, otherwise sub frame's pointers in
+         * dst may be invalid. */
+        if (force_copy && sd_src->type != AV_FRAME_DATA_SUB_FRAME) {
             sd_dst = av_frame_new_side_data(dst, sd_src->type,
                                             sd_src->size);
             if (!sd_dst) {
diff --git a/libavutil/frame.h b/libavutil/frame.h
index 3b6d746a16..31ae7fc08e 100644
--- a/libavutil/frame.h
+++ b/libavutil/frame.h
@@ -228,6 +228,11 @@ enum AVFrameSideDataType {
      * encoding.
      */
     AV_FRAME_DATA_VIDEO_HINT,
+
+    /**
+     * Sub frame of a target frame, as described by AVFrame.
+     */
+    AV_FRAME_DATA_SUB_FRAME,
 };
 
 enum AVActiveFormatDescription {
diff --git a/libavutil/sub_frame_metadata.c b/libavutil/sub_frame_metadata.c
new file mode 100644
index 0000000000..82ea32383f
--- /dev/null
+++ b/libavutil/sub_frame_metadata.c
@@ -0,0 +1,61 @@
+/*
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "sub_frame_metadata.h"
+
+static void sub_frame_free(void *opaque, uint8_t *data)
+{
+    AVFrame *frame = (AVFrame*)data;
+
+    av_frame_free(&frame);
+}
+
+static AVFrame *sub_frame_alloc(size_t *out_size)
+{
+    AVFrame *sub_frame = av_frame_alloc();
+    if (!sub_frame)
+        return NULL;
+
+    *out_size = sizeof(*sub_frame);
+
+    return sub_frame;
+}
+
+AVFrame *av_sub_frame_create_side_data(AVFrame *frame)
+{
+    AVBufferRef *buf;
+    AVFrame *sub_frame;
+    size_t size;
+
+    sub_frame = sub_frame_alloc(&size);
+    if (!sub_frame)
+        return NULL;
+
+    buf = av_buffer_create((uint8_t *)sub_frame, size, &sub_frame_free, NULL, 0);
+    if (!buf) {
+        av_frame_free(&sub_frame);
+        return NULL;
+    }
+
+    if (!av_frame_new_side_data_from_buf(frame, AV_FRAME_DATA_SUB_FRAME, buf)) {
+        av_buffer_unref(&buf);
+        return NULL;
+    }
+
+    return sub_frame;
+}
diff --git a/libavutil/sub_frame_metadata.h b/libavutil/sub_frame_metadata.h
new file mode 100644
index 0000000000..621fb31e42
--- /dev/null
+++ b/libavutil/sub_frame_metadata.h
@@ -0,0 +1,35 @@
+/*
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef AVUTIL_SUB_FRAME_METADATA_H
+#define AVUTIL_SUB_FRAME_METADATA_H
+
+#include "frame.h"
+
+/**
+ * Allocate a AVFrame structure and add it to the input frame as
+ * the side data. The allocated AVFrame will be freed automatically
+ * once the buf of created side data reference count decrease to zero.
+ *
+ * @param frame The frame which side data is added to.
+ *
+ * @return The AVFrame structure to be filled by caller.
+ */
+AVFrame *av_sub_frame_create_side_data(AVFrame *frame);
+
+#endif /* AVUTIL_SUB_FRAME_METADATA_H */
-- 
2.25.1

